{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b9eb00b4-ae7d-4ca0-bb58-a1446ebd1336",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import re\n",
    "import shutil\n",
    "import random\n",
    "import pprint\n",
    "import itertools\n",
    "import functools\n",
    "import collections\n",
    "\n",
    "import pysam\n",
    "import pyranges as pr\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import scipy.stats\n",
    "import sklearn.cluster\n",
    "\n",
    "import sys\n",
    "sys.path.append('/home/users/pjh/scripts/python_genome_package_dev')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "890bcb37-27ab-4015-9776-188dd47fd5b9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from handygenome import common\n",
    "from handygenome.common import ChromDict, Interval\n",
    "from handygenome.variant.vcfspec import Vcfspec\n",
    "\n",
    "from handygenome.sv.breakends import Breakends\n",
    "from handygenome.variant.variantplus import VariantPlus, VariantPlusList\n",
    "from handygenome.igvhandle import IGVHandle\n",
    "\n",
    "import handygenome.variant.filter as libfilter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a56e414-2f9d-48a2-a5d2-7a24a72ebcfa",
   "metadata": {},
   "source": [
    "# Setups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "667ec54c-107e-4142-8a44-6fb05f002a78",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "SAMPLEIDS = [\n",
    "    '14', '6', '87', 'F13', 'F2', 'F33', 'F37', 'FF1', 'FF104', 'FF115', 'FF13', 'FF18', 'FF20', 'FF21', 'FF23', 'FF24', 'FF27', \n",
    "    'FF3', 'FF31', 'FF34', 'FF37', 'FF39', 'FF4', 'FF43', 'FF53', 'FF56', 'FF57', 'FF58', 'FF6', 'FF62', 'FF67', 'FF71', 'FF76', \n",
    "    'FF77', 'FF78', 'FF79', 'FF80', 'FF85', \n",
    "    #'IO04',  # swapped sample\n",
    "    'IO05', 'IO06', 'IO07', 'IO08', 'IO09', 'IO12', 'IO13', 'IO15', 'IO16', 'IO17', \n",
    "    'IO18', 'IO22', 'IO23', 'IO24', 'IO25', 'SC126', 'SC134', 'SC31', 'SC81', 'SC88', 'SC97', 'SC98', \n",
    "]\n",
    "SAMPLEIDS_LU = [f'LU-{x}' for x in SAMPLEIDS]\n",
    "\n",
    "BAM_TOPDIR = '/home/users/team_projects/Lung_Cancer_Panel_data/03_Data_from_YTKim/02_BAM/'\n",
    "\n",
    "sampletype_filename_map = {\n",
    "    'tumor': ['tumor'], \n",
    "    'normal': ['normal'], \n",
    "    'panel': ['panel'], \n",
    "    'rna': ['RNA', 'RNA_GFP'],\n",
    "}\n",
    "BAM_PATHS = {sampletype: dict() for sampletype in sampletype_filename_map.keys()}\n",
    "\n",
    "for top, dirs, files in os.walk(BAM_TOPDIR):\n",
    "    for f in files:\n",
    "        for sampleid in SAMPLEIDS_LU:\n",
    "            for sampletype, fname_portions in sampletype_filename_map.items():\n",
    "                if f in (f'{sampleid}.{fname_portion}.bam' for fname_portion in fname_portions):\n",
    "                    BAM_PATHS[sampletype][sampleid] = os.path.join(top, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "32f2be98-3887-41d1-ab8a-7f5c913c97f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from handygenome import ucscdata as ucscdata\n",
    "cytoband_gr = ucscdata.get_cytoband_gr('hg19')\n",
    "centromere_gr = cytoband_gr[cytoband_gr.Stain == 'acen']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "defaa4dc-cff3-4c54-a7ad-28bb90697cb4",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8ae1b35-8e61-4189-ac14-c452ab27d29c",
   "metadata": {},
   "source": [
    "### Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cf1ecac8-1160-4ba7-902e-8b8e61f27995",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "topdir = '/home/users/team_projects/Lung_Cancer_Panel_data/03_Data_from_YTKim/23_binned_depths/results'\n",
    "all_normal_100bin_files = dict()\n",
    "\n",
    "for top, dirs, files in os.walk(topdir):\n",
    "    for f in files:\n",
    "        if f.endswith('normal.regions.bed.gz'):\n",
    "            sid = f.split('.')[0]\n",
    "            abspath = os.path.join(top, f)\n",
    "            all_normal_100bin_files[sid] = abspath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "59b72e93-e479-4f95-bf3e-70e4581491eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import multiprocessing\n",
    "import itertools\n",
    "\n",
    "nproc = 18\n",
    "\n",
    "keys = list(all_normal_100bin_files.keys())\n",
    "kwargs = dict(sep='\\t', names=['Chromosome', 'Start', 'End', 'depth'], dtype={'Chromosome': str, 'Start': int, 'End': int, 'depth': float})\n",
    "\n",
    "def worker(x, y):\n",
    "    return pd.read_csv(x, **y)\n",
    "\n",
    "with multiprocessing.Pool(nproc) as pool:\n",
    "    dflist = pool.starmap(\n",
    "        worker,\n",
    "        zip(\n",
    "            (all_normal_100bin_files[key] for key in keys),\n",
    "            itertools.repeat(kwargs),\n",
    "        )\n",
    "    )\n",
    "\n",
    "normal_100bin_dfs = dict(zip(keys, dflist))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b970c938-194d-42b3-92a2-a67bed67a3c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "normal_100bin_df_list = list(normal_100bin_dfs.items())\n",
    "merged_dfs = pd.concat(\n",
    "    (\n",
    "        [normal_100bin_df_list[0][1]]\n",
    "        + [x[1].iloc[:, 3] for x in normal_100bin_df_list[1:]]\n",
    "    ),\n",
    "    names=(\n",
    "        ['Chromosome', 'Start', 'End'] \n",
    "        + [x[0] for x in normal_100bin_df_list]\n",
    "    ),\n",
    "    axis=1,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "74d54ba5-05a7-4fd9-aa5a-9b3b32eb2cf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_dfs.columns = (['Chromosome', 'Start', 'End'] + [x[0] for x in normal_100bin_df_list])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "db61b7ce-35e4-4c83-83fd-2c2ee015e7fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_gr = pr.PyRanges(merged_dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "faaf9a8e-464f-466a-8b5f-45cf4dc39c6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "coordcols = merged_dfs.iloc[:, :3]\n",
    "datacols = merged_dfs.iloc[:, 3:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "077af07c-0ab7-434d-aec2-dedc30b55312",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8df982a4-33ad-4b4c-b4f3-dd18c5209307",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "26b7b48f-1abc-4715-b1d5-c20f524a1de7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_ranks(data):\n",
    "    argsort = scipy.stats.rankdata(data, method='max')\n",
    "    return argsort / len(argsort)\n",
    "\n",
    "\n",
    "def pick_row_view(df, igv):\n",
    "    row = df.sample(n=1)\n",
    "    chrom, start0, end0 = row.iloc[0, :3]\n",
    "    locus = (chrom, start0 + 1, end0)\n",
    "    print(locus)\n",
    "    \n",
    "    row_data = row.iloc[0, 3:]\n",
    "    lt9995_samples = row_data.loc[row_data < 0.9995].index.to_list()\n",
    "    gt9995_samples = row_data.loc[row_data > 0.9995].index.to_list()\n",
    "    print('lt', lt9995_samples)\n",
    "    \n",
    "    igv.new()\n",
    "    igv.goto([locus])\n",
    "    igv.load([BAM_PATHS['normal'][x] for x in lt9995_samples])\n",
    "    igv.load([BAM_PATHS['normal'][x] for x in np.random.choice(gt9995_samples, size=3)])\n",
    "    igv.cmd('squish')\n",
    "    \n",
    "    return row\n",
    "\n",
    "\n",
    "def pick_row_view_mean(row, igv, means):\n",
    "    # row = df.sample(n=1)\n",
    "    chrom, start0, end0 = row.iloc[:3]\n",
    "    locus = (chrom, start0 + 1, end0)\n",
    "    print(locus)\n",
    "    \n",
    "    row_data = row.loc[[x for x in row.index if x.startswith('LU-')]]\n",
    "    # lt_samples = row_data.index[row_data < (means * 4)].to_list()\n",
    "    # gt_samples = row_data.index[row_data > (means * 4)].to_list()\n",
    "\n",
    "    sorted_mean_ratios = (row_data / means).sort_values()\n",
    "    # print(sorted_mean_ratios)\n",
    "    least_samples = sorted_mean_ratios.index[:5]\n",
    "    greatest_samples = sorted_mean_ratios.index[-5:]\n",
    "    \n",
    "    print('least samples')\n",
    "    for sample in least_samples:\n",
    "        print(f'sample={sample}, mean={means[sample]}, current_depth={row_data[sample]}, ratio={sorted_mean_ratios[sample]}')\n",
    "    print('greatest samples')\n",
    "    for sample in greatest_samples:\n",
    "        print(f'sample={sample}, mean={means[sample]}, current_depth={row_data[sample]}, ratio={sorted_mean_ratios[sample]}')\n",
    "    \n",
    "    # print('lt', lt_samples)\n",
    "    # print('gt', gt_samples)\n",
    "    \n",
    "    igv.new()\n",
    "    igv.goto([locus])\n",
    "    # igv.load([BAM_PATHS['normal'][x] for x in np.random.choice(lt_samples, replace=False, size=min(3, len(lt_samples)))])\n",
    "    # igv.load([BAM_PATHS['normal'][x] for x in np.random.choice(gt_samples, replace=False, size=min(3, len(gt_samples)))])\n",
    "    igv.load([BAM_PATHS['normal'][x] for x in least_samples])\n",
    "    igv.load([BAM_PATHS['normal'][x] for x in greatest_samples])\n",
    "    igv.cmd('squish')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86aaaf8f-9a82-4ee3-a87b-8adbcefbf565",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "45a4f5c6-2bab-40d4-87c2-9c1bdb46e898",
   "metadata": {},
   "source": [
    "### Create candidate blacklist region"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80c8f978-6ba3-4883-837d-410cacc9282f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "##### Rank-based"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "a5c51707-fadb-435a-abb9-650c2406302b",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[187], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# quantile_datacols = datacols.apply(get_ranks, axis=0)\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m datacols_rank \u001b[38;5;241m=\u001b[39m \u001b[43mdatacols\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrank\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmax\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpct\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/tools/miniconda/221104/miniconda3/envs/genome_v7/lib/python3.10/site-packages/pandas/core/generic.py:9169\u001b[0m, in \u001b[0;36mNDFrame.rank\u001b[0;34m(self, axis, method, numeric_only, na_option, ascending, pct)\u001b[0m\n\u001b[1;32m   9167\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m numeric_only \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   9168\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 9169\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mranker\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m   9170\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   9171\u001b[0m         numeric_only \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m~/tools/miniconda/221104/miniconda3/envs/genome_v7/lib/python3.10/site-packages/pandas/core/generic.py:9153\u001b[0m, in \u001b[0;36mNDFrame.rank.<locals>.ranker\u001b[0;34m(data)\u001b[0m\n\u001b[1;32m   9145\u001b[0m     ranks \u001b[38;5;241m=\u001b[39m values\u001b[38;5;241m.\u001b[39m_rank(\n\u001b[1;32m   9146\u001b[0m         axis\u001b[38;5;241m=\u001b[39maxis,\n\u001b[1;32m   9147\u001b[0m         method\u001b[38;5;241m=\u001b[39mmethod,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   9150\u001b[0m         pct\u001b[38;5;241m=\u001b[39mpct,\n\u001b[1;32m   9151\u001b[0m     )\n\u001b[1;32m   9152\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 9153\u001b[0m     ranks \u001b[38;5;241m=\u001b[39m \u001b[43malgos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrank\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   9154\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9155\u001b[0m \u001b[43m        \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9156\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9157\u001b[0m \u001b[43m        \u001b[49m\u001b[43mascending\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mascending\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9158\u001b[0m \u001b[43m        \u001b[49m\u001b[43mna_option\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mna_option\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9159\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpct\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpct\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9160\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   9162\u001b[0m ranks_obj \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_constructor(ranks, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mdata\u001b[38;5;241m.\u001b[39m_construct_axes_dict())\n\u001b[1;32m   9163\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m ranks_obj\u001b[38;5;241m.\u001b[39m__finalize__(\u001b[38;5;28mself\u001b[39m, method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrank\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/tools/miniconda/221104/miniconda3/envs/genome_v7/lib/python3.10/site-packages/pandas/core/algorithms.py:1147\u001b[0m, in \u001b[0;36mrank\u001b[0;34m(values, axis, method, na_option, ascending, pct)\u001b[0m\n\u001b[1;32m   1138\u001b[0m     ranks \u001b[38;5;241m=\u001b[39m algos\u001b[38;5;241m.\u001b[39mrank_1d(\n\u001b[1;32m   1139\u001b[0m         values,\n\u001b[1;32m   1140\u001b[0m         is_datetimelike\u001b[38;5;241m=\u001b[39mis_datetimelike,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1144\u001b[0m         pct\u001b[38;5;241m=\u001b[39mpct,\n\u001b[1;32m   1145\u001b[0m     )\n\u001b[1;32m   1146\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m values\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m2\u001b[39m:\n\u001b[0;32m-> 1147\u001b[0m     ranks \u001b[38;5;241m=\u001b[39m \u001b[43malgos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrank_2d\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1148\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1149\u001b[0m \u001b[43m        \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1150\u001b[0m \u001b[43m        \u001b[49m\u001b[43mis_datetimelike\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_datetimelike\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1151\u001b[0m \u001b[43m        \u001b[49m\u001b[43mties_method\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1152\u001b[0m \u001b[43m        \u001b[49m\u001b[43mascending\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mascending\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1153\u001b[0m \u001b[43m        \u001b[49m\u001b[43mna_option\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mna_option\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1154\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpct\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpct\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1155\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1156\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1157\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mArray with ndim > 2 are not supported.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# quantile_datacols = datacols.apply(get_ranks, axis=0)\n",
    "datacols_rank = datacols.rank(axis=0, method='max', pct=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "6b1ee72f-bdd8-435e-bbea-9ce52f579464",
   "metadata": {},
   "outputs": [],
   "source": [
    "pctrank_df = pd.concat([coordcols, quantile_datacols], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "6008cb7c-9522-437f-8f4e-bf6d81dbed00",
   "metadata": {},
   "outputs": [],
   "source": [
    "datacols_rank_gt9995 = quantile_datacols > 0.9995"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "id": "7a194dfa-6dd4-4a7b-a732-dd2bdb3a3cf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "rank_gt9995_samplefraction = datacols_rank_gt9995.mean(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "84afd2dd-90b5-4bff-98c7-4997fa17aa17",
   "metadata": {},
   "outputs": [],
   "source": [
    "pctrank_gt9995_morethan95pct = rank_gt9995_samplefraction > 0.95"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23cafcb5-dc97-4d86-9960-238c80ca6bd9",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Mean-based"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "32947fee-d761-4883-868e-e8d24b75a04c",
   "metadata": {},
   "outputs": [],
   "source": [
    "means = datacols.mean(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "e351ff01-f1b4-4224-aa4c-f9303f300bb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "datacols_gtmean = dict()\n",
    "gtmean_samplefracs = dict()\n",
    "gtmean_samplecounts = dict()\n",
    "\n",
    "def add_meancompare_data(fraction):\n",
    "    datacols_gtmean[fraction] = datacols > (fraction * means)\n",
    "    gtmean_samplefracs[fraction] = datacols_gtmean[fraction].mean(axis=1)\n",
    "    gtmean_samplecounts[fraction] = datacols_gtmean[fraction].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "626c4d5c-4bb7-4120-8001-f63f89f4f095",
   "metadata": {},
   "outputs": [],
   "source": [
    "add_meancompare_data(3)\n",
    "add_meancompare_data(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "c23e93ff-a8e3-4003-a62a-907c34b67d58",
   "metadata": {},
   "outputs": [],
   "source": [
    "igv = IGVHandle(port=60387)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "44403d42-7471-4ebc-9743-0c23241ccf59",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_dfs_annot = merged_dfs.assign(**{'gtmean_samplecounts': gtmean_samplecounts[4]})\n",
    "# subdf = merged_dfs.assign(**{'gtmean_samplecounts': gtmean_samplecounts[4]})\n",
    "merged_dfs_annot.sort_values('gtmean_samplecounts', ascending=True, inplace=True)\n",
    "subdf = merged_dfs_annot.loc[merged_dfs_annot['gtmean_samplecounts'] >= 5, :]\n",
    "\n",
    "# subdf = merged_dfs.loc[gtmean_samplefracs[4].between(0.5, 0.6).array, :]\n",
    "# subdf = merged_dfs.loc[(gtmean_samplefracs[3] > 0.5).array, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "df0202dd-a992-4cb2-8438-8bbdb0fb0c20",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(76007, 64)\n",
      "(31018086, 63)\n"
     ]
    }
   ],
   "source": [
    "print(subdf.shape)\n",
    "print(merged_dfs.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "e72003c9-3046-4eb2-a424-2188f12203d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>LU-FF80</th>\n",
       "      <th>LU-IO18</th>\n",
       "      <th>LU-FF76</th>\n",
       "      <th>LU-14</th>\n",
       "      <th>LU-IO05</th>\n",
       "      <th>LU-IO23</th>\n",
       "      <th>LU-SC134</th>\n",
       "      <th>...</th>\n",
       "      <th>LU-IO24</th>\n",
       "      <th>LU-SC81</th>\n",
       "      <th>LU-FF104</th>\n",
       "      <th>LU-FF57</th>\n",
       "      <th>LU-IO12</th>\n",
       "      <th>LU-FF85</th>\n",
       "      <th>LU-FF27</th>\n",
       "      <th>LU-SC98</th>\n",
       "      <th>LU-FF53</th>\n",
       "      <th>gtmean_samplecounts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16089684</th>\n",
       "      <td>9</td>\n",
       "      <td>69808300</td>\n",
       "      <td>69808400</td>\n",
       "      <td>151.40</td>\n",
       "      <td>142.82</td>\n",
       "      <td>163.17</td>\n",
       "      <td>154.71</td>\n",
       "      <td>158.36</td>\n",
       "      <td>119.42</td>\n",
       "      <td>159.19</td>\n",
       "      <td>...</td>\n",
       "      <td>165.99</td>\n",
       "      <td>150.59</td>\n",
       "      <td>94.05</td>\n",
       "      <td>106.21</td>\n",
       "      <td>141.48</td>\n",
       "      <td>138.30</td>\n",
       "      <td>109.24</td>\n",
       "      <td>179.40</td>\n",
       "      <td>73.19</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14052360</th>\n",
       "      <td>8</td>\n",
       "      <td>12440000</td>\n",
       "      <td>12440100</td>\n",
       "      <td>163.55</td>\n",
       "      <td>87.32</td>\n",
       "      <td>160.99</td>\n",
       "      <td>82.44</td>\n",
       "      <td>164.61</td>\n",
       "      <td>149.41</td>\n",
       "      <td>95.53</td>\n",
       "      <td>...</td>\n",
       "      <td>92.23</td>\n",
       "      <td>95.80</td>\n",
       "      <td>74.90</td>\n",
       "      <td>58.42</td>\n",
       "      <td>90.30</td>\n",
       "      <td>117.43</td>\n",
       "      <td>66.63</td>\n",
       "      <td>135.15</td>\n",
       "      <td>87.04</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31000368</th>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>112000</td>\n",
       "      <td>112100</td>\n",
       "      <td>106.39</td>\n",
       "      <td>92.12</td>\n",
       "      <td>109.07</td>\n",
       "      <td>149.02</td>\n",
       "      <td>125.35</td>\n",
       "      <td>87.57</td>\n",
       "      <td>91.73</td>\n",
       "      <td>...</td>\n",
       "      <td>152.74</td>\n",
       "      <td>127.05</td>\n",
       "      <td>63.55</td>\n",
       "      <td>89.65</td>\n",
       "      <td>155.39</td>\n",
       "      <td>101.93</td>\n",
       "      <td>82.35</td>\n",
       "      <td>25.26</td>\n",
       "      <td>70.54</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30975211</th>\n",
       "      <td>GL000214.1</td>\n",
       "      <td>103700</td>\n",
       "      <td>103800</td>\n",
       "      <td>97.75</td>\n",
       "      <td>131.08</td>\n",
       "      <td>136.24</td>\n",
       "      <td>96.33</td>\n",
       "      <td>137.36</td>\n",
       "      <td>62.83</td>\n",
       "      <td>174.39</td>\n",
       "      <td>...</td>\n",
       "      <td>137.04</td>\n",
       "      <td>142.48</td>\n",
       "      <td>76.06</td>\n",
       "      <td>69.71</td>\n",
       "      <td>75.91</td>\n",
       "      <td>146.06</td>\n",
       "      <td>58.72</td>\n",
       "      <td>144.77</td>\n",
       "      <td>93.92</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22193738</th>\n",
       "      <td>14</td>\n",
       "      <td>19437000</td>\n",
       "      <td>19437100</td>\n",
       "      <td>96.95</td>\n",
       "      <td>98.25</td>\n",
       "      <td>160.11</td>\n",
       "      <td>104.92</td>\n",
       "      <td>143.91</td>\n",
       "      <td>146.11</td>\n",
       "      <td>120.63</td>\n",
       "      <td>...</td>\n",
       "      <td>134.97</td>\n",
       "      <td>71.19</td>\n",
       "      <td>96.70</td>\n",
       "      <td>45.39</td>\n",
       "      <td>121.87</td>\n",
       "      <td>94.35</td>\n",
       "      <td>75.41</td>\n",
       "      <td>93.04</td>\n",
       "      <td>93.89</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Chromosome     Start       End  LU-FF80  LU-IO18  LU-FF76   LU-14  \\\n",
       "16089684           9  69808300  69808400   151.40   142.82   163.17  154.71   \n",
       "14052360           8  12440000  12440100   163.55    87.32   160.99   82.44   \n",
       "31000368  GL000195.1    112000    112100   106.39    92.12   109.07  149.02   \n",
       "30975211  GL000214.1    103700    103800    97.75   131.08   136.24   96.33   \n",
       "22193738          14  19437000  19437100    96.95    98.25   160.11  104.92   \n",
       "\n",
       "          LU-IO05  LU-IO23  LU-SC134  ...  LU-IO24  LU-SC81  LU-FF104  \\\n",
       "16089684   158.36   119.42    159.19  ...   165.99   150.59     94.05   \n",
       "14052360   164.61   149.41     95.53  ...    92.23    95.80     74.90   \n",
       "31000368   125.35    87.57     91.73  ...   152.74   127.05     63.55   \n",
       "30975211   137.36    62.83    174.39  ...   137.04   142.48     76.06   \n",
       "22193738   143.91   146.11    120.63  ...   134.97    71.19     96.70   \n",
       "\n",
       "          LU-FF57  LU-IO12  LU-FF85  LU-FF27  LU-SC98  LU-FF53  \\\n",
       "16089684   106.21   141.48   138.30   109.24   179.40    73.19   \n",
       "14052360    58.42    90.30   117.43    66.63   135.15    87.04   \n",
       "31000368    89.65   155.39   101.93    82.35    25.26    70.54   \n",
       "30975211    69.71    75.91   146.06    58.72   144.77    93.92   \n",
       "22193738    45.39   121.87    94.35    75.41    93.04    93.89   \n",
       "\n",
       "          gtmean_samplecounts  \n",
       "16089684                    5  \n",
       "14052360                    5  \n",
       "31000368                    5  \n",
       "30975211                    5  \n",
       "22193738                    5  \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "d6188451-f7e8-4237-8b3e-e31e6d7af119",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome                    9\n",
      "Start                  69808300\n",
      "End                    69808400\n",
      "LU-FF80                   151.4\n",
      "LU-IO18                  142.82\n",
      "                         ...   \n",
      "LU-FF85                   138.3\n",
      "LU-FF27                  109.24\n",
      "LU-SC98                   179.4\n",
      "LU-FF53                   73.19\n",
      "gtmean_samplecounts           5\n",
      "Name: 16089684, Length: 64, dtype: object\n",
      "('9', 69808301, 69808400)\n",
      "least samples\n",
      "sample=LU-IO06, mean=41.282814418014084, current_depth=26.4, ratio=0.6394912840167251\n",
      "sample=LU-FF53, mean=31.707044289579997, current_depth=73.19, ratio=2.3083198588792047\n",
      "sample=LU-IO23, mean=47.7968633280596, current_depth=119.42, ratio=2.498490312645545\n",
      "sample=LU-FF24, mean=36.186269118926326, current_depth=97.66, ratio=2.6988137317787584\n",
      "sample=LU-IO22, mean=61.57973562134055, current_depth=172.47, ratio=2.8007590201512698\n",
      "greatest samples\n",
      "sample=LU-FF6, mean=31.188080685249222, current_depth=126.34, ratio=4.050906539425301\n",
      "sample=LU-14, mean=36.9762601215948, current_depth=154.71, ratio=4.184035905503775\n",
      "sample=LU-FF67, mean=33.43857195540688, current_depth=140.77, ratio=4.2098089651594135\n",
      "sample=LU-6, mean=33.33318137876073, current_depth=146.12, ratio=4.383619983333031\n",
      "sample=LU-FF58, mean=30.877867418382905, current_depth=138.55, ratio=4.4870326736850785\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "row = subdf.iloc[0, :]\n",
    "print(row)\n",
    "pick_row_view_mean(row, igv, means)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "161689e6-e99a-498d-991d-fba38b5216c5",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "##### Readstats_annotator --mq-limits --depth-limits test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "id": "b91f5db0-e58b-4747-8bc6-af9e6a91527c",
   "metadata": {},
   "outputs": [],
   "source": [
    "bam_path = '/home/users/team_projects/Lung_Cancer_Panel_data/03_Data_from_YTKim/02_BAM/IO22/LU-IO22.normal.bam'\n",
    "bam = pysam.AlignmentFile(bam_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 413,
   "id": "4d657672-ba41-4e3f-93a6-11d8c936ef38",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom = '2'\n",
    "# pos0 = 213_056_734\n",
    "pos0 = 213_055_850\n",
    "\n",
    "start0 = pos0\n",
    "end0 = pos0 + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 414,
   "id": "324de4ce-7bb8-4695-9b68-70253030b1a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37.7 ms ± 1.29 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "\n",
    "pup = bam.pileup(chrom, start0, end0, truncate=True, stepper='nofilter', ignore_overlaps=False, flag_filter=0, ignore_orphans=False, min_base_quality=0, min_mapping_quality=0)\n",
    "pupcol = next(pup)\n",
    "mqs = pupcol.get_mapping_qualities()\n",
    "depth = pupcol.get_num_aligned()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 415,
   "id": "fa274823-2b84-492e-8c58-1fec4fd08dc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8.18 ms ± 82.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "\n",
    "mqs2 = list()\n",
    "for idx, read in enumerate(bam.fetch(chrom, start0, end0)):\n",
    "    mqs2.append(read.mapping_quality)\n",
    "depth2 = idx + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 409,
   "id": "25728259-4a95-4faa-8a81-135f187b625b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 409,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(mqs2) == sorted(mqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 410,
   "id": "d3f30624-c7b6-416d-81bb-37f54583654f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 410,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depth == depth2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90c1a7a2-9bdf-48cd-8a4a-b88d4f929088",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2eadf97-a69c-446f-b231-682f2a535594",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 416,
   "id": "384d28c3-552f-48d9-a716-e0842fcd553e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import handygenome.annotation.readstats as libreadstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 436,
   "id": "8bce03da-32a2-4206-bdee-b07b078959f3",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "111 19.72072072072072\n"
     ]
    }
   ],
   "source": [
    "depth, mqs = libreadstats.get_position_info(bam, '1', 10751)\n",
    "print(depth, np.mean(mqs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed930a21-308a-4f49-a5ae-da1b0913abc8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 424,
   "id": "a167f470-cf70-40d2-9453-2cfc0f177ac3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# vcfspec = Vcfspec('2', 213_056_736, 'T', ('G',), refver='hg19')\n",
    "vcfspec = Vcfspec('2', 213_056_543, 'T', ('C',), refver='hg19')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 425,
   "id": "e666f4a9-6a00-43fa-9603-7943ea0df8b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "readstats = libreadstats.ReadStats.from_bam(vcfspec, bam, common.DEFAULT_FASTAS['hg19'], common.DEFAULT_CHROMDICTS['hg19'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 426,
   "id": "6a34aa2f-e0a2-4f49-a366-56dbba9f4508",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<ReadStats(\n",
       "    {'rppcounts': {\u001b[38;5;93mNone\u001b[0m: 157, -1: 0, 0: 25, 1: 21, 'softclip_overlap': 157},\n",
       "     'mean_BQs': {\u001b[38;5;93mNone\u001b[0m: nan, -1: nan, 0: 31.76, 1: 30.09090909090909},\n",
       "     'median_BQs': {\u001b[38;5;93mNone\u001b[0m: nan, -1: nan, 0: 32.0, 1: 31.0},\n",
       "     'mean_MQs': {\u001b[38;5;93mNone\u001b[0m: 14.872611464968152, -1: nan, 0: 60.2, 1: 59.5},\n",
       "     'median_MQs': {\u001b[38;5;93mNone\u001b[0m: 12.0, -1: nan, 0: 60.0, 1: 60.0},\n",
       "     'mean_cliplens': {\u001b[38;5;93mNone\u001b[0m: 261.28662420382165,\n",
       "                       -1: nan,\n",
       "                       0: 7.88,\n",
       "                       1: 7.0476190476190474},\n",
       "     'median_cliplens': {\u001b[38;5;93mNone\u001b[0m: 261.0, -1: nan, 0: 3.0, 1: 0.0},\n",
       "     'mNM': {\u001b[38;5;93mNone\u001b[0m: 0.025477707006369428, -1: 0, 0: 0.84, 1: 1.0},\n",
       "     'recurrent_mNM': {\u001b[38;5;93mNone\u001b[0m: 1, -1: 0, 0: 2, 1: 3},\n",
       "     'pairorient_pvalues': {\u001b[38;5;93mNone\u001b[0m: 0.1101572607611227,\n",
       "                            -1: nan,\n",
       "                            0: 0.30745625495910645,\n",
       "                            1: 0.8238029479980469},\n",
       "     'readorient_pvalues': {\u001b[38;5;93mNone\u001b[0m: 1.0, -1: nan, 0: 1.0, 1: 1.0},\n",
       "     'varpos_uniform_pvalues': {\u001b[38;5;93mNone\u001b[0m: nan,\n",
       "                                -1: nan,\n",
       "                                0: 0.0014470989262063139,\n",
       "                                1: 0.10350162166306942},\n",
       "     'mean_varpos_fractions': {\u001b[38;5;93mNone\u001b[0m: nan,\n",
       "                               -1: nan,\n",
       "                               0: 0.6908609271523178,\n",
       "                               1: 0.5743527995183625},\n",
       "     'median_varpos_fractions': {\u001b[38;5;93mNone\u001b[0m: nan,\n",
       "                                 -1: nan,\n",
       "                                 0: 0.7483443708609272,\n",
       "                                 1: 0.6788079470198676}}\n",
       ")>"
      ]
     },
     "execution_count": 426,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "readstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 427,
   "id": "7c010347-0a3a-4f11-8fcb-e0973439531d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 427,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "readstats.is_invalid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bd52378-562d-4df0-9363-552ad2e630d1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2e85a9fa-5597-488c-b6c4-c0eae634e026",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "##### Readstats_annotator long-running regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "674dc1e2-ee6b-41c4-870b-b741c516e8b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import re\n",
    "import pickle\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "sys.path.append('/home/users/pjh/scripts/python_genome_package_dev/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "1aff117b-865b-43f3-9c33-6fe7f0dc065e",
   "metadata": {},
   "outputs": [],
   "source": [
    "LOGDIR = '/home/users/team_projects/Lung_Cancer_Panel_data/03_Data_from_YTKim/06_haplotypecaller/merged_vcfs/all_sample_merged.vcf.gz_readstats_annotator_5adyj0tk/logs'\n",
    "PAT = re.compile(r'\\[.+\\] Processing (.+)')\n",
    "\n",
    "def get_region_from_logfile(fname):\n",
    "    fname = os.path.join(LOGDIR, fname)\n",
    "    coords = list()\n",
    "    with open(fname) as infile:\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            if line == '':\n",
    "                continue\n",
    "                \n",
    "            mat = PAT.fullmatch(line)\n",
    "            if mat is None:\n",
    "                continue\n",
    "                \n",
    "            variantinfo = mat.group(1).split()\n",
    "            chrom, pos = variantinfo[0], int(variantinfo[1])\n",
    "            coords.append((chrom, pos))\n",
    "            \n",
    "    locus = (coords[0], coords[-1])\n",
    "    print(locus[0][0], f'{locus[0][1]:,}')\n",
    "    print(locus[1][0], f'{locus[1][1]:,}')\n",
    "    return locus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7067ea5b-37ce-43bb-a89a-279104564236",
   "metadata": {},
   "outputs": [],
   "source": [
    "import handygenome.ucscdata as ucscdata\n",
    "cytoband_gr = ucscdata.get_cytoband_gr('hg19')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "c9d8aea3-889b-4d19-8957-3e4fad3ec564",
   "metadata": {},
   "outputs": [],
   "source": [
    "from handygenome.igvhandle import IGVHandle\n",
    "igv = IGVHandle(60387)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "6ea6f688-ab50-4f27-856f-b830e0cf81df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OK\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "igv.load([\n",
    "    BAM_PATHS['normal']['LU-14'],\n",
    "    BAM_PATHS['normal']['LU-6'],\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "d6fb87b7-376b-454d-bd90-f4928226ccb1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OK\n"
     ]
    }
   ],
   "source": [
    "igv.cmd('squish')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2a055810-64bd-4b9a-9ecb-414e6e8f8203",
   "metadata": {},
   "outputs": [],
   "source": [
    "pickle_path = '/home/users/team_projects/Lung_Cancer_Panel_data/03_Data_from_YTKim/06_haplotypecaller/workspace/timespent.output.pickle'\n",
    "with open(pickle_path, 'rb') as infile:\n",
    "    data = pickle.load(infile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b0141c21-b091-4491-aeaf-94ff10e8ae86",
   "metadata": {},
   "outputs": [],
   "source": [
    "ser = pd.Series(data)\n",
    "ser_secperline = ser * 60 / 1565"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a537540e-5f0a-4075-a593-09a97c9c046a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([  18.03333333,   42.78333333,   44.1       ,   45.26166667,\n",
       "         46.42666667,   47.63333333,   48.98333333,   50.5       ,\n",
       "         52.65      ,   57.96833333, 2276.16666667])"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.quantile(ser, np.linspace(0, 1, 11, endpoint=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dc9ce944-2e42-4920-8cd6-f12ceb3331e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 0.6913738 ,  1.64025559,  1.69073482,  1.73527157,  1.7799361 ,\n",
       "        1.82619808,  1.87795527,  1.93610224,  2.01853035,  2.22242812,\n",
       "       87.26517572])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.quantile(ser_secperline, np.linspace(0, 1, 11, endpoint=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "bce49db9-bc7c-4ab7-911c-c44b5232de51",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 5726.sbatch.success 2276.1666666666665\n",
      "1 5721.sbatch.success 2084.516666666667\n",
      "2 4275.sbatch.success 1599.45\n",
      "3 5720.sbatch.success 1270.7166666666667\n",
      "4 1087.sbatch.success 1137.5166666666667\n",
      "5 6225.sbatch.success 1117.6166666666666\n",
      "6 2375.sbatch.success 1090.6\n",
      "7 2379.sbatch.success 938.6833333333333\n",
      "8 5717.sbatch.success 896.1333333333333\n",
      "9 2380.sbatch.success 885.1166666666667\n",
      "10 2376.sbatch.success 879.4833333333333\n",
      "11 2377.sbatch.success 855.35\n",
      "12 5719.sbatch.success 847.0\n",
      "13 2373.sbatch.success 724.8\n",
      "14 9828.sbatch.success 719.65\n",
      "15 4791.sbatch.success 683.0166666666667\n",
      "16 9827.sbatch.success 682.65\n",
      "17 8827.sbatch.success 682.3\n",
      "18 1085.sbatch.success 655.0666666666667\n",
      "19 1066.sbatch.success 652.3833333333333\n",
      "20 8527.sbatch.success 646.8833333333333\n",
      "21 8025.sbatch.success 636.4666666666667\n",
      "22 2390.sbatch.success 615.0333333333333\n",
      "23 5725.sbatch.success 577.6833333333333\n",
      "24 2374.sbatch.success 565.8666666666667\n",
      "25 5724.sbatch.success 560.1166666666667\n",
      "26 0874.sbatch.success 557.05\n",
      "27 2378.sbatch.success 536.9666666666667\n",
      "28 1086.sbatch.success 497.4\n",
      "29 5723.sbatch.success 471.56666666666666\n",
      "30 9931.sbatch.success 459.6666666666667\n",
      "31 8284.sbatch.success 458.0833333333333\n",
      "32 8282.sbatch.success 452.26666666666665\n",
      "33 9905.sbatch.success 444.76666666666665\n",
      "34 8463.sbatch.success 430.68333333333334\n",
      "35 8283.sbatch.success 426.56666666666666\n",
      "36 5714.sbatch.success 417.55\n",
      "37 5716.sbatch.success 391.23333333333335\n",
      "38 5712.sbatch.success 391.1166666666667\n",
      "39 9843.sbatch.success 390.3333333333333\n",
      "40 4796.sbatch.success 389.56666666666666\n",
      "41 9842.sbatch.success 364.96666666666664\n",
      "42 8027.sbatch.success 343.3666666666667\n",
      "43 5715.sbatch.success 333.23333333333335\n",
      "44 9927.sbatch.success 333.1333333333333\n",
      "45 2391.sbatch.success 330.3833333333333\n",
      "46 2269.sbatch.success 328.25\n",
      "47 9912.sbatch.success 316.25\n",
      "48 9914.sbatch.success 307.98333333333335\n",
      "49 6224.sbatch.success 303.98333333333335\n",
      "50 9917.sbatch.success 301.1166666666667\n",
      "51 9191.sbatch.success 293.35\n",
      "52 2392.sbatch.success 293.1166666666667\n",
      "53 9921.sbatch.success 288.56666666666666\n",
      "54 2384.sbatch.success 286.6166666666667\n",
      "55 9915.sbatch.success 283.9166666666667\n",
      "56 8011.sbatch.success 271.6666666666667\n",
      "57 9841.sbatch.success 271.2\n",
      "58 8026.sbatch.success 268.3833333333333\n",
      "59 5713.sbatch.success 263.23333333333335\n",
      "60 9962.sbatch.success 263.1333333333333\n",
      "61 9829.sbatch.success 247.43333333333334\n",
      "62 9190.sbatch.success 236.65\n",
      "63 9832.sbatch.success 235.26666666666668\n",
      "64 9985.sbatch.success 226.46666666666667\n",
      "65 6615.sbatch.success 223.63333333333333\n",
      "66 9605.sbatch.success 217.28333333333333\n",
      "67 9346.sbatch.success 208.15\n",
      "68 1084.sbatch.success 206.93333333333334\n",
      "69 9862.sbatch.success 205.31666666666666\n",
      "70 5718.sbatch.success 204.15\n",
      "71 9189.sbatch.success 203.81666666666666\n",
      "72 9930.sbatch.success 203.25\n",
      "73 9857.sbatch.success 200.35\n",
      "74 9604.sbatch.success 200.13333333333333\n",
      "75 9831.sbatch.success 198.51666666666668\n",
      "76 9928.sbatch.success 198.05\n",
      "77 9838.sbatch.success 195.91666666666666\n",
      "78 3672.sbatch.success 194.23333333333332\n",
      "79 9913.sbatch.success 193.85\n",
      "80 0396.sbatch.success 192.58333333333334\n",
      "81 9890.sbatch.success 191.25\n",
      "82 9984.sbatch.success 190.53333333333333\n",
      "83 9926.sbatch.success 189.66666666666666\n",
      "84 9992.sbatch.success 188.51666666666668\n",
      "85 9993.sbatch.success 187.68333333333334\n",
      "86 5284.sbatch.success 187.55\n",
      "87 9844.sbatch.success 184.03333333333333\n",
      "88 9991.sbatch.success 184.03333333333333\n",
      "89 9937.sbatch.success 181.41666666666666\n",
      "90 9951.sbatch.success 180.45\n",
      "91 9989.sbatch.success 180.0\n",
      "92 9918.sbatch.success 177.85\n",
      "93 9911.sbatch.success 177.5\n",
      "94 9994.sbatch.success 172.5\n",
      "95 9856.sbatch.success 171.5\n",
      "96 8016.sbatch.success 169.58333333333334\n",
      "97 9858.sbatch.success 169.2\n",
      "98 9830.sbatch.success 168.93333333333334\n",
      "99 5285.sbatch.success 167.83333333333334\n",
      "100 8828.sbatch.success 166.26666666666668\n",
      "101 8145.sbatch.success 159.4\n",
      "102 7256.sbatch.success 159.33333333333334\n",
      "103 9199.sbatch.success 158.13333333333333\n",
      "104 9983.sbatch.success 157.83333333333334\n",
      "105 9916.sbatch.success 153.33333333333334\n",
      "106 0392.sbatch.success 152.36666666666667\n",
      "107 6614.sbatch.success 149.95\n",
      "108 3663.sbatch.success 149.46666666666667\n",
      "109 3662.sbatch.success 148.96666666666667\n",
      "110 3664.sbatch.success 143.48333333333332\n",
      "111 1212.sbatch.success 142.98333333333332\n",
      "112 9986.sbatch.success 141.8\n",
      "113 9979.sbatch.success 141.8\n",
      "114 0397.sbatch.success 141.15\n",
      "115 9854.sbatch.success 140.93333333333334\n",
      "116 8028.sbatch.success 139.45\n",
      "117 9192.sbatch.success 139.3\n",
      "118 9859.sbatch.success 139.13333333333333\n",
      "119 9834.sbatch.success 138.81666666666666\n",
      "120 4257.sbatch.success 138.75\n",
      "121 9183.sbatch.success 138.58333333333334\n",
      "122 2848.sbatch.success 138.5\n",
      "123 8018.sbatch.success 136.3\n",
      "124 9935.sbatch.success 135.93333333333334\n",
      "125 8526.sbatch.success 134.93333333333334\n",
      "126 9839.sbatch.success 134.8\n",
      "127 9920.sbatch.success 133.5\n",
      "128 9885.sbatch.success 132.93333333333334\n",
      "129 4795.sbatch.success 132.93333333333334\n",
      "130 9997.sbatch.success 132.61666666666667\n",
      "131 9861.sbatch.failure 132.5\n",
      "132 5289.sbatch.success 131.06666666666666\n",
      "133 9181.sbatch.success 130.4\n",
      "134 4276.sbatch.success 129.93333333333334\n",
      "135 9929.sbatch.success 129.43333333333334\n",
      "136 8274.sbatch.success 128.58333333333334\n",
      "137 3666.sbatch.success 128.05\n",
      "138 9188.sbatch.success 127.9\n",
      "139 4268.sbatch.success 127.63333333333334\n",
      "140 8830.sbatch.success 127.15\n",
      "141 4246.sbatch.success 127.1\n",
      "142 3665.sbatch.success 126.58333333333333\n",
      "143 9826.sbatch.success 125.18333333333334\n",
      "144 7257.sbatch.success 124.58333333333333\n",
      "145 9987.sbatch.success 122.43333333333334\n",
      "146 3661.sbatch.success 122.38333333333334\n",
      "147 8014.sbatch.success 121.65\n",
      "148 7569.sbatch.success 121.36666666666666\n",
      "149 1083.sbatch.success 120.56666666666666\n",
      "150 9200.sbatch.success 120.53333333333333\n",
      "151 9866.sbatch.success 119.73333333333333\n",
      "152 4270.sbatch.success 119.41666666666667\n",
      "153 5317.sbatch.success 118.88333333333334\n",
      "154 9058.sbatch.success 118.55\n",
      "155 9988.sbatch.success 118.23333333333333\n",
      "156 4682.sbatch.success 118.18333333333334\n",
      "157 9970.sbatch.success 117.48333333333333\n",
      "158 8275.sbatch.success 116.91666666666667\n",
      "159 9932.sbatch.success 116.41666666666667\n",
      "160 2386.sbatch.success 115.4\n",
      "161 5722.sbatch.success 115.25\n",
      "162 9919.sbatch.success 114.76666666666667\n",
      "163 4250.sbatch.success 114.7\n",
      "164 5708.sbatch.success 114.63333333333334\n",
      "165 9950.sbatch.success 114.55\n",
      "166 3987.sbatch.success 114.51666666666667\n",
      "167 9347.sbatch.success 114.06666666666666\n",
      "168 0400.sbatch.success 113.6\n",
      "169 5295.sbatch.success 113.26666666666667\n",
      "170 9050.sbatch.success 112.9\n",
      "171 5707.sbatch.success 112.73333333333333\n",
      "172 5314.sbatch.success 112.38333333333334\n",
      "173 0401.sbatch.success 112.33333333333333\n",
      "174 0013.sbatch.success 112.0\n",
      "175 9815.sbatch.success 111.71666666666667\n",
      "176 2847.sbatch.success 111.56666666666666\n",
      "177 9980.sbatch.success 111.56666666666666\n",
      "178 9968.sbatch.success 111.48333333333333\n",
      "179 1855.sbatch.success 111.3\n",
      "180 9603.sbatch.success 110.86666666666666\n",
      "181 9171.sbatch.success 110.73333333333333\n",
      "182 8012.sbatch.success 110.68333333333334\n",
      "183 2382.sbatch.success 110.5\n",
      "184 8279.sbatch.success 110.4\n",
      "185 9837.sbatch.success 110.26666666666667\n",
      "186 9180.sbatch.success 110.16666666666667\n",
      "187 9981.sbatch.success 109.93333333333334\n",
      "188 9899.sbatch.success 109.9\n",
      "189 9936.sbatch.success 109.7\n",
      "190 8003.sbatch.success 109.01666666666667\n",
      "191 9910.sbatch.success 108.88333333333334\n",
      "192 2846.sbatch.success 108.55\n",
      "193 8024.sbatch.success 108.23333333333333\n",
      "194 9996.sbatch.success 107.9\n",
      "195 9990.sbatch.success 107.9\n",
      "196 9051.sbatch.success 107.7\n",
      "197 5731.sbatch.success 107.68333333333334\n",
      "198 6220.sbatch.success 107.66666666666667\n",
      "199 7996.sbatch.success 107.4\n",
      "200 0425.sbatch.success 107.03333333333333\n",
      "201 8829.sbatch.success 106.95\n",
      "202 7576.sbatch.success 106.86666666666666\n",
      "203 9982.sbatch.success 106.55\n",
      "204 7577.sbatch.success 106.55\n",
      "205 8277.sbatch.success 105.7\n",
      "206 5320.sbatch.success 105.5\n",
      "207 0424.sbatch.success 105.23333333333333\n",
      "208 0072.sbatch.success 105.15\n",
      "209 9179.sbatch.success 105.15\n",
      "210 2388.sbatch.success 104.66666666666667\n",
      "211 9922.sbatch.success 104.45\n",
      "212 5706.sbatch.success 104.2\n",
      "213 9607.sbatch.success 104.16666666666667\n",
      "214 3150.sbatch.success 104.06666666666666\n",
      "215 7990.sbatch.success 103.93333333333334\n",
      "216 9201.sbatch.success 103.46666666666667\n",
      "217 1057.sbatch.success 103.45\n",
      "218 9819.sbatch.success 103.43333333333334\n",
      "219 9974.sbatch.success 103.21666666666667\n",
      "220 9954.sbatch.success 103.2\n",
      "221 9185.sbatch.success 103.06666666666666\n",
      "222 0074.sbatch.success 102.98333333333333\n",
      "223 4271.sbatch.success 102.36666666666666\n",
      "224 2205.sbatch.success 102.31666666666666\n",
      "225 9978.sbatch.success 102.2\n",
      "226 0075.sbatch.success 102.16666666666667\n",
      "227 8525.sbatch.success 102.13333333333334\n",
      "228 4273.sbatch.success 101.55\n",
      "229 5732.sbatch.success 101.35\n",
      "230 1213.sbatch.success 101.28333333333333\n",
      "231 1056.sbatch.success 101.25\n",
      "232 9977.sbatch.success 101.25\n",
      "233 1065.sbatch.success 101.21666666666667\n",
      "234 1070.sbatch.success 101.21666666666667\n",
      "235 7997.sbatch.success 101.21666666666667\n",
      "236 0071.sbatch.success 101.15\n",
      "237 9944.sbatch.success 101.01666666666667\n",
      "238 6617.sbatch.success 100.8\n",
      "239 7995.sbatch.success 100.56666666666666\n",
      "240 9949.sbatch.success 100.38333333333334\n",
      "241 0426.sbatch.success 100.21666666666667\n",
      "242 0399.sbatch.success 100.21666666666667\n"
     ]
    }
   ],
   "source": [
    "for idx, (key, val) in enumerate(ser[ser>100].sort_values(ascending=False).to_dict().items()):\n",
    "    print(idx, key, val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "1ec78486-7681-4a3d-9b81-d54530f771dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0874.sbatch.success\n",
      "2 32,877,384\n",
      "2 33,185,302\n"
     ]
    }
   ],
   "source": [
    "idx = 26\n",
    "fname = ser[ser>100].sort_values(ascending=False).index[idx]\n",
    "print(fname)\n",
    "locus = get_region_from_logfile(fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "ddb104d0-a776-4351-93f1-690b87988d61",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\t90,500,000\t93,300,000\n",
      "2\t93,300,000\t96,800,000\n"
     ]
    }
   ],
   "source": [
    "subgr = cytoband_gr[cytoband_gr.Stain == 'acen'][locus[0][0]]\n",
    "\n",
    "for idx, row in subgr.df.iterrows():\n",
    "    print(row[0], f'{row[1]:,}', f'{row[2]:,}', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "60677356-d26c-4bd1-8cdc-f992a199e6c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OK\n"
     ]
    }
   ],
   "source": [
    "igv.goto(\n",
    "    [\n",
    "        (locus[0][0], locus[0][1], locus[0][1])\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "01015451-e2b6-4a93-b6b7-652db235c552",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OK\n"
     ]
    }
   ],
   "source": [
    "igv.goto(\n",
    "    [\n",
    "        ('Y', 13_400_000, 13_400_000)\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afe49c3d-5f40-4014-8480-592a7723c653",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### Check cancer gene overlap"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e70899d-0233-42a3-8014-52502f40ea46",
   "metadata": {},
   "source": [
    "##### Get cancer gene coordinates & intersect with candidate blacklist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "adad31ea-da2a-4da4-8da4-b533d6477466",
   "metadata": {},
   "outputs": [],
   "source": [
    "import handygenome.analysis.feature as analysisfeat\n",
    "import handygenome.annotation.oncokb as liboncokb\n",
    "\n",
    "ONCOKB_TOKEN = '864a64ee-cfa0-4b99-aa25-305fb4bcda39'\n",
    "oncokb_cancergenes = liboncokb.get_cancerGeneList(ONCOKB_TOKEN)\n",
    "oncokb_cancergenes_coords = analysisfeat.get_gene_coords_gr(oncokb_cancergenes['hugoSymbol'].to_list(), 'hg19').sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "66a3230b-d1b0-4093-9307-8ab0b8d4630d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1682670</td>\n",
       "      <td>1711896</td>\n",
       "      <td>NADK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1716728</td>\n",
       "      <td>1822495</td>\n",
       "      <td>GNB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2487077</td>\n",
       "      <td>2496821</td>\n",
       "      <td>TNFRSF14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2985731</td>\n",
       "      <td>3355185</td>\n",
       "      <td>PRDM16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>6241328</td>\n",
       "      <td>6269449</td>\n",
       "      <td>RPL22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1025</th>\n",
       "      <td>X</td>\n",
       "      <td>152783133</td>\n",
       "      <td>152848397</td>\n",
       "      <td>ATP2B3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1026</th>\n",
       "      <td>X</td>\n",
       "      <td>152907945</td>\n",
       "      <td>152916781</td>\n",
       "      <td>DUSP9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1027</th>\n",
       "      <td>X</td>\n",
       "      <td>153618314</td>\n",
       "      <td>153637504</td>\n",
       "      <td>RPL10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1028</th>\n",
       "      <td>X</td>\n",
       "      <td>153656977</td>\n",
       "      <td>153664862</td>\n",
       "      <td>ATP6AP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1029</th>\n",
       "      <td>X</td>\n",
       "      <td>154289896</td>\n",
       "      <td>154376212</td>\n",
       "      <td>MTCP1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1030 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------------+\n",
       "| Chromosome   | Start     | End       | gene_name   |\n",
       "| (category)   | (int32)   | (int32)   | (object)    |\n",
       "|--------------+-----------+-----------+-------------|\n",
       "| 1            | 1682670   | 1711896   | NADK        |\n",
       "| 1            | 1716728   | 1822495   | GNB1        |\n",
       "| 1            | 2487077   | 2496821   | TNFRSF14    |\n",
       "| 1            | 2985731   | 3355185   | PRDM16      |\n",
       "| ...          | ...       | ...       | ...         |\n",
       "| X            | 152907945 | 152916781 | DUSP9       |\n",
       "| X            | 153618314 | 153637504 | RPL10       |\n",
       "| X            | 153656977 | 153664862 | ATP6AP1     |\n",
       "| X            | 154289896 | 154376212 | MTCP1       |\n",
       "+--------------+-----------+-----------+-------------+\n",
       "Unstranded PyRanges object has 1,030 rows and 4 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oncokb_cancergenes_coords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "3b8bca41-df9e-4c95-b5fb-b7379c4d37c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "subdf_gr = pr.PyRanges(subdf)\n",
    "subdf_cgeneannot_gr = subdf_gr.join(oncokb_cancergenes_coords)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a31e4042-426e-48fe-aa32-0c2043b38560",
   "metadata": {},
   "source": [
    "##### Load panel gene list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "49cd6b49-c0f0-43a1-a2c5-d862206af171",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "curated_drivers = {\n",
    "    'tcga_adc': [ # The Cancer Genome Atlas Research Network. Comprehensive molecular profiling of lung adenocarcinoma. Nature 511, 543–550 (2014).\n",
    "        'RIT1',\n",
    "        'U2AF1',\n",
    "        'CDKN2A',\n",
    "        'RB1',\n",
    "        'SMARCA4',\n",
    "        'PIK3CA',\n",
    "        'ARID1A',\n",
    "        'MET',\n",
    "        'MGA',\n",
    "        'RBM10',\n",
    "        'SETD2',\n",
    "        'BRAF',\n",
    "        'NF1',\n",
    "        'EGFR',\n",
    "        'STK11',\n",
    "        'KEAP1',\n",
    "        'KRAS',\n",
    "        'TP53',\n",
    "    ],\n",
    "    'tcga_sqcc': [ # The Cancer Genome Atlas Research Network. Comprehensive genomic characterization of squamous cell lung cancers. Nature 489, 519–525 (2012).\n",
    "        'RB1',\n",
    "        'NOTCH1',\n",
    "        'NFE2L2',\n",
    "        'HLA-A',\n",
    "        #'MLL2', - synonym of KMT2D\n",
    "        'KEAP1',\n",
    "        'PIK3CA',\n",
    "        'PTEN',\n",
    "        'CDKN2A',\n",
    "        'TP53',        \n",
    "    ],\n",
    "    'natgen2016_adc': [ # Cancer Genome Atlas Research Network et al. Distinct patterns of somatic genome alterations in lung adenocarcinomas and squamous cell carcinomas. Nat Genet 48, 607–616 (2016).\n",
    "        'TP53',\n",
    "        'KRAS',\n",
    "        'KEAP1',\n",
    "        'EGFR',\n",
    "        'STK11',\n",
    "        'SMARCA4',\n",
    "        'RBM10',\n",
    "        'RB1',\n",
    "        'NF1',\n",
    "        'ARID1A',\n",
    "        'BRAF',\n",
    "        'ERBB2',\n",
    "        'SETD2',\n",
    "        'MGA',\n",
    "        #'FTSJD1', - synonym of CMTR2\n",
    "        'CMTR2',\n",
    "        'MET',\n",
    "        'ATM',\n",
    "        'CDKN2A',\n",
    "        'U2AF1',\n",
    "        'RIT1',\n",
    "        'DOT1L',\n",
    "        'ARID2',\n",
    "        'SMAD4',\n",
    "        'PTPRU',\n",
    "        'CTNNB1',\n",
    "        'ARHGEF12',\n",
    "        'APC',\n",
    "        'KLHL5',\n",
    "        'PIK3CA',\n",
    "        'PPP3CA',\n",
    "        'ATF7IP',\n",
    "        'KARS',\n",
    "        'RAF1',\n",
    "        #'MLL3', - synonym of KMT2C\n",
    "        'FANCM',\n",
    "        'STIM1',\n",
    "        'NRAS',\n",
    "        'MAP2K1',\n",
    "    ],\n",
    "    'natgen2016_sqcc': [ # Cancer Genome Atlas Research Network et al. Distinct patterns of somatic genome alterations in lung adenocarcinomas and squamous cell carcinomas. Nat Genet 48, 607–616 (2016).\n",
    "        'TP53',\n",
    "        'CDKN2A',\n",
    "        'NFE2L2',\n",
    "        'PTEN',\n",
    "        #'MLL2', - synonym of KMT2D\n",
    "        'RB1',\n",
    "        'FAT1',\n",
    "        'NOTCH1',\n",
    "        'RASA1',\n",
    "        'NF1',\n",
    "        'ARID1A',\n",
    "        'KDM6A',\n",
    "        'PIK3CA',\n",
    "        'CUL3',\n",
    "        'HRAS',\n",
    "        'IRF6',\n",
    "        'FBXW7',\n",
    "        'ARHGAP35',\n",
    "        'PASK',\n",
    "        'NSD1',\n",
    "    ],\n",
    "    'natrev2019_conventional': [ # Skoulidis, F. & Heymach, J. V. Co-occurring genomic alterations in non-small-cell lung cancer biology and therapy. Nat Rev Cancer 19, 495–509 (2019).\n",
    "        'KRAS',\n",
    "        'EGFR',\n",
    "        'BRAF',\n",
    "        'NF1',\n",
    "        'ERBB2',\n",
    "        'MET',\n",
    "        'ALK',\n",
    "        'ROS1',\n",
    "        'RET',\n",
    "        'MET',\n",
    "        'ERBB2',\n",
    "        'MAP2K1',\n",
    "        'NRAS',\n",
    "        'HRAS',\n",
    "        'RIT1',\n",
    "        'FGFR2',\n",
    "        'FGFR1',\n",
    "    ],\n",
    "    'natrev2019_KRAS_excl': [ # Skoulidis, F. & Heymach, J. V. Co-occurring genomic alterations in non-small-cell lung cancer biology and therapy. Nat Rev Cancer 19, 495–509 (2019).\n",
    "        'EGFR',\n",
    "        'EML4',\n",
    "        'ERBB2',\n",
    "        'TP53',\n",
    "        'ROS1',\n",
    "    ],\n",
    "    'natrev2019_KRAS_co': [ # Skoulidis, F. & Heymach, J. V. Co-occurring genomic alterations in non-small-cell lung cancer biology and therapy. Nat Rev Cancer 19, 495–509 (2019).\n",
    "        #'LKB1', - synonym of STK11\n",
    "        'KEAP1', \n",
    "        'ATM',\n",
    "        'RBM10',\n",
    "        'PTPRD',\n",
    "        'U2AF1',\n",
    "        'POLE',\n",
    "        'NTRK3',\n",
    "    ],\n",
    "    'natrev2019_EGFR_excl': [ # Skoulidis, F. & Heymach, J. V. Co-occurring genomic alterations in non-small-cell lung cancer biology and therapy. Nat Rev Cancer 19, 495–509 (2019).\n",
    "        'KRAS',\n",
    "        'KEAP1',\n",
    "        #'LKB1', - synonym of STK11\n",
    "        'ALK',\n",
    "        'EML4',\n",
    "        'MET',\n",
    "        'DDR2',\n",
    "        'BRCA2',\n",
    "        'PTPRT',\n",
    "        'PTPRD',\n",
    "        'NF1',\n",
    "        #'EPHA',\n",
    "        'EPHA1',\n",
    "        'EPHA2',\n",
    "        'EPHA3',\n",
    "        'EPHA4',\n",
    "        'EPHA5',\n",
    "        'EPHA6',\n",
    "        'EPHA7',\n",
    "        'EPHA8',\n",
    "        'EPHA10',\n",
    "        'RET',\n",
    "        #'PAK5', - synonym of PAK6\n",
    "        'PAK6',\n",
    "        'ERBB4',\n",
    "        'KMT2C',\n",
    "        'TBX3',\n",
    "        'FAT1',\n",
    "    ],\n",
    "    'natrev2019_EGFR_co': [ # Skoulidis, F. & Heymach, J. V. Co-occurring genomic alterations in non-small-cell lung cancer biology and therapy. Nat Rev Cancer 19, 495–509 (2019).\n",
    "        'TP53',\n",
    "        'RB1',\n",
    "        'PIK3CA',\n",
    "        'CTNNB1',\n",
    "    ],\n",
    "    'manually_added_221113': [\n",
    "        # from benchmark ppt\n",
    "        'PIK3CA',\n",
    "        'PIK3R1',\n",
    "        'PIK3C3',\n",
    "        'PIK3C2G',\n",
    "        'PIK3CG',\n",
    "        'PIK3R2',\n",
    "        'PIK3CA',\n",
    "        'ARID2',\n",
    "        'CCND2',\n",
    "        'DNMT3A',\n",
    "        'FGFR4',\n",
    "        'MTOR',\n",
    "        'NCOR1',\n",
    "        'NKX2-1',\n",
    "        'SF3B1',\n",
    "        'SMARCA2',\n",
    "        'SMARCB1',\n",
    "        # from TCGA paper fig 3\n",
    "        'RIT1',\n",
    "    ],\n",
    "}\n",
    "\n",
    "panel_genes_exon = [\n",
    "    'AKT1',\n",
    "    'AKT3',\n",
    "    'ALK',\n",
    "    'ARAF',\n",
    "    'ARID1A',\n",
    "    'ATM',\n",
    "    'ATR',\n",
    "    'AXL',\n",
    "    'BRAF',\n",
    "    'BRCA1',\n",
    "    'BRCA2',\n",
    "    'CBL',\n",
    "    'CCND1',\n",
    "    'CD274',\n",
    "    'CDK11B',\n",
    "    'CDKN2A',\n",
    "    'CHEK2',\n",
    "    'CREBBP',\n",
    "    'CTNNB1',\n",
    "    'DDR2',\n",
    "    'EGFR',\n",
    "    'EP300',\n",
    "    'ERBB2',\n",
    "    'ERBB3',\n",
    "    'ERBB4',\n",
    "    'FBXW7',\n",
    "    'FGFR1',\n",
    "    'FGFR2',\n",
    "    'FGFR3',\n",
    "    'HRAS',\n",
    "    'IDH1',\n",
    "    'IDH2',\n",
    "    'IGF1R',\n",
    "    'JAK2',\n",
    "    'JAK3',\n",
    "    'KDR',\n",
    "    'KEAP1',\n",
    "    'KIT',\n",
    "    'KMT2D',\n",
    "    'KRAS',\n",
    "    'MAP2K1',\n",
    "    'MAP2K2',\n",
    "    'MAP2K4',\n",
    "    'MDM2',\n",
    "    'MET',\n",
    "    'MYC',\n",
    "    'MYCN',\n",
    "    'NEK2',\n",
    "    'NF1',\n",
    "    'NFE2L2',\n",
    "    'NOTCH1',\n",
    "    'NRAS',\n",
    "    'NRG1',\n",
    "    'NTRK1',\n",
    "    'NTRK2',\n",
    "    'NTRK3',\n",
    "    'PDGFRA',\n",
    "    'PIK3CA',\n",
    "    'PTEN',\n",
    "    'RAF1',\n",
    "    'RB1',\n",
    "    'RBM10',\n",
    "    'RET',\n",
    "    'RIT1',\n",
    "    'ROS1',\n",
    "    'SDK1',\n",
    "    'SETD2',\n",
    "    'SMAD4',\n",
    "    'SMARCA4',\n",
    "    'SMG1',\n",
    "    'SOX2',\n",
    "    'STK11',\n",
    "    'TERT',\n",
    "    'TP53',\n",
    "    'TP63',\n",
    "]\n",
    "\n",
    "panel_genes_intron = [\n",
    "    'ALK',\n",
    "    'AXL',\n",
    "    'BRAF',\n",
    "    'EGFR',\n",
    "    'ERBB4',\n",
    "    'FGFR1',\n",
    "    'FGFR2',\n",
    "    'FGFR3',\n",
    "    'MET',\n",
    "    'NRG1',\n",
    "    'NTRK1',\n",
    "    'NTRK2',\n",
    "    'NTRK3',\n",
    "    'PDGFRA',\n",
    "    'RET',\n",
    "    'ROS1',    \n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "ac2d9da5-24fd-4fce-bdc1-3ee2ba163634",
   "metadata": {},
   "outputs": [],
   "source": [
    "panel_genes_all = set(itertools.chain(panel_genes_intron, panel_genes_exon))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "d53d9335-fa1f-4a53-9f38-dc10819eeadf",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ABI1',\n",
       " 'ACSL3',\n",
       " 'ACTB',\n",
       " 'AFF1',\n",
       " 'AKT3',\n",
       " 'APC',\n",
       " 'ARHGAP35',\n",
       " 'ARID1B',\n",
       " 'ASXL1',\n",
       " 'ATRX',\n",
       " 'BCL2',\n",
       " 'BCOR',\n",
       " 'BRSK1',\n",
       " 'CAMTA1',\n",
       " 'CBFA2T3',\n",
       " 'CCDC6',\n",
       " 'CCND3',\n",
       " 'CDC42',\n",
       " 'CDK12',\n",
       " 'CLTCL1',\n",
       " 'CREBBP',\n",
       " 'CRKL',\n",
       " 'CTNNA1',\n",
       " 'CUL4A',\n",
       " 'CUX1',\n",
       " 'DDX6',\n",
       " 'DICER1',\n",
       " 'DNM2',\n",
       " 'DNMT1',\n",
       " 'DNMT3A',\n",
       " 'EBF1',\n",
       " 'EP300',\n",
       " 'EP400',\n",
       " 'EPHA5',\n",
       " 'EPHB1',\n",
       " 'ERBB4',\n",
       " 'ERC1',\n",
       " 'ESR1',\n",
       " 'EWSR1',\n",
       " 'EXT1',\n",
       " 'FLT4',\n",
       " 'FOXP1',\n",
       " 'GAB2',\n",
       " 'GAS7',\n",
       " 'GNA13',\n",
       " 'GPC3',\n",
       " 'GPHN',\n",
       " 'GRIN2A',\n",
       " 'GRM3',\n",
       " 'INPP5D',\n",
       " 'INSR',\n",
       " 'KDM4C',\n",
       " 'KDM6A',\n",
       " 'KLF3',\n",
       " 'KMT2C',\n",
       " 'LIFR',\n",
       " 'LPP',\n",
       " 'LRP1B',\n",
       " 'MAP4K4',\n",
       " 'MAPKAP1',\n",
       " 'MDC1',\n",
       " 'MECOM',\n",
       " 'MEF2C',\n",
       " 'MGAM',\n",
       " 'MIB1',\n",
       " 'MKNK1',\n",
       " 'MSH2',\n",
       " 'MSH3',\n",
       " 'MSH6',\n",
       " 'MUC1',\n",
       " 'MYH9',\n",
       " 'NBEAP1',\n",
       " 'NCOA1',\n",
       " 'NCOA4',\n",
       " 'NCOR1',\n",
       " 'NCOR2',\n",
       " 'NEGR1',\n",
       " 'NF1',\n",
       " 'NOTCH2',\n",
       " 'NRG1',\n",
       " 'NSD1',\n",
       " 'NT5C2',\n",
       " 'NUP214',\n",
       " 'NUP98',\n",
       " 'PASK',\n",
       " 'PAX3',\n",
       " 'PBRM1',\n",
       " 'PCM1',\n",
       " 'PDE4DIP',\n",
       " 'PIK3CB',\n",
       " 'PLCG2',\n",
       " 'POLD1',\n",
       " 'PPARG',\n",
       " 'PRCC',\n",
       " 'PRDM16',\n",
       " 'PTCH1',\n",
       " 'PTPRD',\n",
       " 'RAD51',\n",
       " 'RAD51B',\n",
       " 'REST',\n",
       " 'REV3L',\n",
       " 'ROBO1',\n",
       " 'RPTOR',\n",
       " 'RSPO3',\n",
       " 'RTEL1',\n",
       " 'SETD3',\n",
       " 'SETDB2',\n",
       " 'SHOC2',\n",
       " 'SMARCA1',\n",
       " 'SMC1A',\n",
       " 'SMG1',\n",
       " 'SMYD3',\n",
       " 'SND1',\n",
       " 'SRC',\n",
       " 'STAG1',\n",
       " 'STAG2',\n",
       " 'SUZ12',\n",
       " 'TAF1',\n",
       " 'TCF7L2',\n",
       " 'TERT',\n",
       " 'TET1',\n",
       " 'TLE4',\n",
       " 'TSHR',\n",
       " 'TUSC3',\n",
       " 'U2AF2',\n",
       " 'WWP1',\n",
       " 'YWHAE',\n",
       " 'YY1',\n",
       " 'ZBTB16',\n",
       " 'ZBTB20',\n",
       " 'ZNF521']"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(subdf_cgeneannot_gr.gene_name.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "9a4001b8-0e72-4fda-ad29-d718ab694432",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AKT3', 'CREBBP', 'EP300', 'ERBB4', 'NF1', 'NRG1', 'SMG1', 'TERT'}"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "panel_genes_all.intersection(subdf_cgeneannot_gr.gene_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "5fec0701-261f-4c09-8bb0-ffc003f29b30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>LU-FF80</th>\n",
       "      <th>LU-IO18</th>\n",
       "      <th>LU-FF76</th>\n",
       "      <th>LU-14</th>\n",
       "      <th>LU-IO05</th>\n",
       "      <th>LU-IO23</th>\n",
       "      <th>LU-SC134</th>\n",
       "      <th>...</th>\n",
       "      <th>LU-FF57</th>\n",
       "      <th>LU-IO12</th>\n",
       "      <th>LU-FF85</th>\n",
       "      <th>LU-FF27</th>\n",
       "      <th>LU-SC98</th>\n",
       "      <th>LU-FF53</th>\n",
       "      <th>gtmean_samplecounts</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14</td>\n",
       "      <td>81561000</td>\n",
       "      <td>81561100</td>\n",
       "      <td>267.11</td>\n",
       "      <td>442.89</td>\n",
       "      <td>30.33</td>\n",
       "      <td>35.34</td>\n",
       "      <td>299.47</td>\n",
       "      <td>365.64</td>\n",
       "      <td>185.21</td>\n",
       "      <td>...</td>\n",
       "      <td>33.16</td>\n",
       "      <td>355.57</td>\n",
       "      <td>35.97</td>\n",
       "      <td>26.45</td>\n",
       "      <td>250.38</td>\n",
       "      <td>24.99</td>\n",
       "      <td>16</td>\n",
       "      <td>81421332</td>\n",
       "      <td>81612646</td>\n",
       "      <td>TSHR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14</td>\n",
       "      <td>81469100</td>\n",
       "      <td>81469200</td>\n",
       "      <td>4319.40</td>\n",
       "      <td>8988.26</td>\n",
       "      <td>13.77</td>\n",
       "      <td>21.35</td>\n",
       "      <td>5390.95</td>\n",
       "      <td>6153.41</td>\n",
       "      <td>2932.21</td>\n",
       "      <td>...</td>\n",
       "      <td>27.63</td>\n",
       "      <td>6728.94</td>\n",
       "      <td>15.57</td>\n",
       "      <td>20.83</td>\n",
       "      <td>4128.51</td>\n",
       "      <td>13.29</td>\n",
       "      <td>19</td>\n",
       "      <td>81421332</td>\n",
       "      <td>81612646</td>\n",
       "      <td>TSHR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 67 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------------+-------------+-------+\n",
       "|   Chromosome |     Start |       End |     LU-FF80 |     LU-IO18 | +62   |\n",
       "|   (category) |   (int32) |   (int32) |   (float64) |   (float64) | ...   |\n",
       "|--------------+-----------+-----------+-------------+-------------+-------|\n",
       "|           14 |  81561000 |  81561100 |      267.11 |      442.89 | ...   |\n",
       "|           14 |  81469100 |  81469200 |     4319.4  |     8988.26 | ...   |\n",
       "+--------------+-----------+-----------+-------------+-------------+-------+\n",
       "Unstranded PyRanges object has 2 rows and 67 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "62 hidden columns: LU-FF76, LU-14, LU-IO05, LU-IO23, LU-SC134, LU-FF4, LU-FF67, ... (+ 55 more.)"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subdf_cgeneannot_gr[subdf_cgeneannot_gr.gene_name == 'TSHR']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "eb7a008f-5bcb-457e-a0ae-5b4690f99203",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome                   14\n",
      "Start                  81469100\n",
      "End                    81469200\n",
      "LU-FF80                  4319.4\n",
      "LU-IO18                 8988.26\n",
      "                         ...   \n",
      "LU-FF53                   13.29\n",
      "gtmean_samplecounts          19\n",
      "Start_b                81421332\n",
      "End_b                  81612646\n",
      "gene_name                  TSHR\n",
      "Name: 1, Length: 67, dtype: object\n",
      "('14', 81469101, 81469200)\n",
      "least samples\n",
      "sample=LU-FF56, mean=28.942476457122417, current_depth=8.72, ratio=0.30128727971562735\n",
      "sample=LU-FF76, mean=45.61566637380509, current_depth=13.77, ratio=0.30186997351215844\n",
      "sample=LU-FF78, mean=32.53089656208966, current_depth=12.42, ratio=0.3817908915081616\n",
      "sample=LU-6, mean=33.33318137876073, current_depth=12.76, ratio=0.3828017450542669\n",
      "sample=LU-FF39, mean=25.121992081007214, current_depth=9.87, ratio=0.3928828561116353\n",
      "greatest samples\n",
      "sample=LU-IO24, mean=49.76233501544866, current_depth=5982.02, ratio=120.21180272474933\n",
      "sample=LU-IO25, mean=48.90238395173711, current_depth=6203.0, ratio=126.84453187643948\n",
      "sample=LU-IO23, mean=47.7968633280596, current_depth=6153.41, ratio=128.74087485124954\n",
      "sample=LU-IO12, mean=46.86154232630598, current_depth=6728.94, ratio=143.5919448221548\n",
      "sample=LU-IO18, mean=47.74400304905996, current_depth=8988.26, ratio=188.25945513542294\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "row = subdf_cgeneannot_gr[subdf_cgeneannot_gr.gene_name == 'TSHR'].df.iloc[1, :]\n",
    "# idx = 6\n",
    "# row = cancergene_overlap_gr.df.iloc[idx:(idx + 1), :]\n",
    "print(row)\n",
    "pick_row_view_mean(row, igv, means)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc558aa5-4afb-4f8f-9d08-e191dc35e04e",
   "metadata": {},
   "source": [
    "### Confirm Final blacklist and save it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "11a62d0e-a945-4e38-b85a-8b5f434112e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "blacklist_gr = pr.PyRanges(subdf).sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "88a2e059-c09d-4f6c-998a-255f9d31d8d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "blacklist_df = blacklist_gr.df.rename(columns={'gtmean_samplecounts': 'gt_4mean_samplecounts'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "ac496044-afb5-4c55-8718-52fd674ad097",
   "metadata": {},
   "outputs": [],
   "source": [
    "blacklist_path = '/home/users/pjh/scripts/python_genome_package_dev/data/custom_blacklist.tsv.gz'\n",
    "blacklist_df.to_csv(blacklist_path, sep='\\t', header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12590577-a840-4f36-bed7-371e8dd2ae66",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79777c76-10e8-4ed5-913f-f43c682b962d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "5b59d866-2bd6-493e-845b-8a1f131b361a",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_dfs_annot_gr = pr.PyRanges(merged_dfs_annot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "605c1d01-6f26-4b00-b573-a9a2859c3a99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>LU-FF80</th>\n",
       "      <th>LU-IO18</th>\n",
       "      <th>LU-FF76</th>\n",
       "      <th>LU-14</th>\n",
       "      <th>LU-IO05</th>\n",
       "      <th>LU-IO23</th>\n",
       "      <th>LU-SC134</th>\n",
       "      <th>...</th>\n",
       "      <th>LU-IO24</th>\n",
       "      <th>LU-SC81</th>\n",
       "      <th>LU-FF104</th>\n",
       "      <th>LU-FF57</th>\n",
       "      <th>LU-IO12</th>\n",
       "      <th>LU-FF85</th>\n",
       "      <th>LU-FF27</th>\n",
       "      <th>LU-SC98</th>\n",
       "      <th>LU-FF53</th>\n",
       "      <th>gtmean_samplecounts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>33141200</td>\n",
       "      <td>33141300</td>\n",
       "      <td>10061.92</td>\n",
       "      <td>18901.03</td>\n",
       "      <td>662.90</td>\n",
       "      <td>476.56</td>\n",
       "      <td>12624.36</td>\n",
       "      <td>12815.05</td>\n",
       "      <td>8621.38</td>\n",
       "      <td>...</td>\n",
       "      <td>13262.95</td>\n",
       "      <td>617.76</td>\n",
       "      <td>454.16</td>\n",
       "      <td>493.94</td>\n",
       "      <td>15653.00</td>\n",
       "      <td>440.86</td>\n",
       "      <td>395.86</td>\n",
       "      <td>9882.05</td>\n",
       "      <td>408.98</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>33141300</td>\n",
       "      <td>33141400</td>\n",
       "      <td>2922557.42</td>\n",
       "      <td>5869068.87</td>\n",
       "      <td>66513.92</td>\n",
       "      <td>47291.85</td>\n",
       "      <td>3747546.08</td>\n",
       "      <td>3914055.12</td>\n",
       "      <td>2341136.90</td>\n",
       "      <td>...</td>\n",
       "      <td>4048134.33</td>\n",
       "      <td>59262.13</td>\n",
       "      <td>45194.47</td>\n",
       "      <td>49687.68</td>\n",
       "      <td>4742511.44</td>\n",
       "      <td>46518.35</td>\n",
       "      <td>38207.37</td>\n",
       "      <td>2901473.37</td>\n",
       "      <td>40231.16</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>33141400</td>\n",
       "      <td>33141500</td>\n",
       "      <td>2308283.28</td>\n",
       "      <td>4532227.15</td>\n",
       "      <td>71166.99</td>\n",
       "      <td>55250.19</td>\n",
       "      <td>3082463.35</td>\n",
       "      <td>2926391.11</td>\n",
       "      <td>2285220.69</td>\n",
       "      <td>...</td>\n",
       "      <td>3215387.23</td>\n",
       "      <td>69190.25</td>\n",
       "      <td>46811.78</td>\n",
       "      <td>52499.79</td>\n",
       "      <td>3897396.19</td>\n",
       "      <td>57191.96</td>\n",
       "      <td>39426.46</td>\n",
       "      <td>2365816.21</td>\n",
       "      <td>41335.07</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>33141500</td>\n",
       "      <td>33141600</td>\n",
       "      <td>5895586.12</td>\n",
       "      <td>11218106.59</td>\n",
       "      <td>148116.46</td>\n",
       "      <td>113737.55</td>\n",
       "      <td>7746268.78</td>\n",
       "      <td>7473889.92</td>\n",
       "      <td>5767356.35</td>\n",
       "      <td>...</td>\n",
       "      <td>7946781.27</td>\n",
       "      <td>144794.17</td>\n",
       "      <td>95392.92</td>\n",
       "      <td>109254.27</td>\n",
       "      <td>9374017.79</td>\n",
       "      <td>113462.85</td>\n",
       "      <td>78539.73</td>\n",
       "      <td>6094810.55</td>\n",
       "      <td>82335.65</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>33141600</td>\n",
       "      <td>33141700</td>\n",
       "      <td>3188920.02</td>\n",
       "      <td>6199340.58</td>\n",
       "      <td>53445.20</td>\n",
       "      <td>41923.58</td>\n",
       "      <td>4140398.21</td>\n",
       "      <td>4108340.56</td>\n",
       "      <td>2875539.71</td>\n",
       "      <td>...</td>\n",
       "      <td>4332996.57</td>\n",
       "      <td>51917.16</td>\n",
       "      <td>34638.30</td>\n",
       "      <td>39518.79</td>\n",
       "      <td>5090919.61</td>\n",
       "      <td>39525.65</td>\n",
       "      <td>28274.88</td>\n",
       "      <td>3227658.74</td>\n",
       "      <td>29720.33</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-----------------+-------+\n",
       "|   Chromosome |     Start |       End |         LU-FF80 | +60   |\n",
       "|   (category) |   (int32) |   (int32) |       (float64) | ...   |\n",
       "|--------------+-----------+-----------+-----------------+-------|\n",
       "|            2 |  33141200 |  33141300 | 10061.9         | ...   |\n",
       "|            2 |  33141300 |  33141400 |     2.92256e+06 | ...   |\n",
       "|            2 |  33141400 |  33141500 |     2.30828e+06 | ...   |\n",
       "|            2 |  33141500 |  33141600 |     5.89559e+06 | ...   |\n",
       "|            2 |  33141600 |  33141700 |     3.18892e+06 | ...   |\n",
       "+--------------+-----------+-----------+-----------------+-------+\n",
       "Unstranded PyRanges object has 5 rows and 64 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "60 hidden columns: LU-IO18, LU-FF76, LU-14, LU-IO05, LU-IO23, LU-SC134, ... (+ 54 more.)"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blacklist_gr['2', 33_140_000:33_142_000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "31b2961b-e2e1-4788-8b19-55a9dfab6446",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>LU-FF80</th>\n",
       "      <th>LU-IO18</th>\n",
       "      <th>LU-FF76</th>\n",
       "      <th>LU-14</th>\n",
       "      <th>LU-IO05</th>\n",
       "      <th>LU-IO23</th>\n",
       "      <th>LU-SC134</th>\n",
       "      <th>...</th>\n",
       "      <th>LU-IO24</th>\n",
       "      <th>LU-SC81</th>\n",
       "      <th>LU-FF104</th>\n",
       "      <th>LU-FF57</th>\n",
       "      <th>LU-IO12</th>\n",
       "      <th>LU-FF85</th>\n",
       "      <th>LU-FF27</th>\n",
       "      <th>LU-SC98</th>\n",
       "      <th>LU-FF53</th>\n",
       "      <th>gtmean_samplecounts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>33140000</td>\n",
       "      <td>33140100</td>\n",
       "      <td>42.00</td>\n",
       "      <td>39.76</td>\n",
       "      <td>37.49</td>\n",
       "      <td>30.77</td>\n",
       "      <td>40.25</td>\n",
       "      <td>43.94</td>\n",
       "      <td>51.30</td>\n",
       "      <td>...</td>\n",
       "      <td>42.40</td>\n",
       "      <td>33.62</td>\n",
       "      <td>38.72</td>\n",
       "      <td>32.53</td>\n",
       "      <td>37.13</td>\n",
       "      <td>52.61</td>\n",
       "      <td>27.80</td>\n",
       "      <td>49.78</td>\n",
       "      <td>25.57</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>33140100</td>\n",
       "      <td>33140200</td>\n",
       "      <td>44.61</td>\n",
       "      <td>30.61</td>\n",
       "      <td>31.30</td>\n",
       "      <td>25.01</td>\n",
       "      <td>38.54</td>\n",
       "      <td>44.71</td>\n",
       "      <td>49.52</td>\n",
       "      <td>...</td>\n",
       "      <td>41.39</td>\n",
       "      <td>35.16</td>\n",
       "      <td>28.90</td>\n",
       "      <td>30.34</td>\n",
       "      <td>44.15</td>\n",
       "      <td>32.02</td>\n",
       "      <td>24.43</td>\n",
       "      <td>41.99</td>\n",
       "      <td>20.96</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>33140200</td>\n",
       "      <td>33140300</td>\n",
       "      <td>47.35</td>\n",
       "      <td>38.67</td>\n",
       "      <td>30.33</td>\n",
       "      <td>22.10</td>\n",
       "      <td>41.72</td>\n",
       "      <td>56.64</td>\n",
       "      <td>36.09</td>\n",
       "      <td>...</td>\n",
       "      <td>45.31</td>\n",
       "      <td>32.66</td>\n",
       "      <td>22.29</td>\n",
       "      <td>32.48</td>\n",
       "      <td>40.57</td>\n",
       "      <td>25.65</td>\n",
       "      <td>25.63</td>\n",
       "      <td>39.84</td>\n",
       "      <td>32.03</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>33140300</td>\n",
       "      <td>33140400</td>\n",
       "      <td>44.05</td>\n",
       "      <td>38.27</td>\n",
       "      <td>33.23</td>\n",
       "      <td>25.90</td>\n",
       "      <td>43.43</td>\n",
       "      <td>55.24</td>\n",
       "      <td>40.42</td>\n",
       "      <td>...</td>\n",
       "      <td>49.04</td>\n",
       "      <td>37.68</td>\n",
       "      <td>27.92</td>\n",
       "      <td>29.32</td>\n",
       "      <td>47.93</td>\n",
       "      <td>28.16</td>\n",
       "      <td>26.31</td>\n",
       "      <td>37.39</td>\n",
       "      <td>32.64</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>33140400</td>\n",
       "      <td>33140500</td>\n",
       "      <td>52.68</td>\n",
       "      <td>33.66</td>\n",
       "      <td>45.54</td>\n",
       "      <td>39.52</td>\n",
       "      <td>49.08</td>\n",
       "      <td>51.61</td>\n",
       "      <td>64.30</td>\n",
       "      <td>...</td>\n",
       "      <td>50.31</td>\n",
       "      <td>53.46</td>\n",
       "      <td>40.37</td>\n",
       "      <td>35.44</td>\n",
       "      <td>57.76</td>\n",
       "      <td>38.94</td>\n",
       "      <td>29.78</td>\n",
       "      <td>45.88</td>\n",
       "      <td>34.90</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>33140500</td>\n",
       "      <td>33140600</td>\n",
       "      <td>53.65</td>\n",
       "      <td>42.15</td>\n",
       "      <td>40.95</td>\n",
       "      <td>26.03</td>\n",
       "      <td>54.46</td>\n",
       "      <td>57.70</td>\n",
       "      <td>64.74</td>\n",
       "      <td>...</td>\n",
       "      <td>39.10</td>\n",
       "      <td>46.77</td>\n",
       "      <td>27.33</td>\n",
       "      <td>38.17</td>\n",
       "      <td>47.43</td>\n",
       "      <td>33.55</td>\n",
       "      <td>33.06</td>\n",
       "      <td>55.66</td>\n",
       "      <td>29.88</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>33140600</td>\n",
       "      <td>33140700</td>\n",
       "      <td>55.44</td>\n",
       "      <td>42.58</td>\n",
       "      <td>37.72</td>\n",
       "      <td>30.41</td>\n",
       "      <td>60.69</td>\n",
       "      <td>50.36</td>\n",
       "      <td>47.17</td>\n",
       "      <td>...</td>\n",
       "      <td>39.90</td>\n",
       "      <td>45.72</td>\n",
       "      <td>28.13</td>\n",
       "      <td>35.36</td>\n",
       "      <td>55.39</td>\n",
       "      <td>35.57</td>\n",
       "      <td>37.54</td>\n",
       "      <td>56.51</td>\n",
       "      <td>31.05</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>33140700</td>\n",
       "      <td>33140800</td>\n",
       "      <td>59.57</td>\n",
       "      <td>44.32</td>\n",
       "      <td>40.14</td>\n",
       "      <td>37.35</td>\n",
       "      <td>58.31</td>\n",
       "      <td>37.27</td>\n",
       "      <td>44.98</td>\n",
       "      <td>...</td>\n",
       "      <td>41.09</td>\n",
       "      <td>48.37</td>\n",
       "      <td>34.42</td>\n",
       "      <td>32.69</td>\n",
       "      <td>50.99</td>\n",
       "      <td>37.95</td>\n",
       "      <td>26.93</td>\n",
       "      <td>49.25</td>\n",
       "      <td>29.51</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>33140800</td>\n",
       "      <td>33140900</td>\n",
       "      <td>51.48</td>\n",
       "      <td>39.11</td>\n",
       "      <td>42.35</td>\n",
       "      <td>33.69</td>\n",
       "      <td>51.15</td>\n",
       "      <td>30.47</td>\n",
       "      <td>42.10</td>\n",
       "      <td>...</td>\n",
       "      <td>33.90</td>\n",
       "      <td>49.19</td>\n",
       "      <td>24.23</td>\n",
       "      <td>43.10</td>\n",
       "      <td>54.08</td>\n",
       "      <td>34.08</td>\n",
       "      <td>24.46</td>\n",
       "      <td>33.38</td>\n",
       "      <td>26.59</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2</td>\n",
       "      <td>33140900</td>\n",
       "      <td>33141000</td>\n",
       "      <td>33.87</td>\n",
       "      <td>29.00</td>\n",
       "      <td>33.91</td>\n",
       "      <td>21.52</td>\n",
       "      <td>48.52</td>\n",
       "      <td>29.47</td>\n",
       "      <td>30.40</td>\n",
       "      <td>...</td>\n",
       "      <td>27.49</td>\n",
       "      <td>36.80</td>\n",
       "      <td>21.40</td>\n",
       "      <td>31.88</td>\n",
       "      <td>36.69</td>\n",
       "      <td>33.67</td>\n",
       "      <td>27.17</td>\n",
       "      <td>33.31</td>\n",
       "      <td>28.83</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2</td>\n",
       "      <td>33141000</td>\n",
       "      <td>33141100</td>\n",
       "      <td>41.08</td>\n",
       "      <td>26.23</td>\n",
       "      <td>33.02</td>\n",
       "      <td>19.93</td>\n",
       "      <td>42.48</td>\n",
       "      <td>29.56</td>\n",
       "      <td>36.60</td>\n",
       "      <td>...</td>\n",
       "      <td>21.88</td>\n",
       "      <td>38.51</td>\n",
       "      <td>19.54</td>\n",
       "      <td>27.08</td>\n",
       "      <td>30.60</td>\n",
       "      <td>24.85</td>\n",
       "      <td>22.69</td>\n",
       "      <td>33.16</td>\n",
       "      <td>27.18</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2</td>\n",
       "      <td>33141100</td>\n",
       "      <td>33141200</td>\n",
       "      <td>21.44</td>\n",
       "      <td>15.36</td>\n",
       "      <td>23.90</td>\n",
       "      <td>8.64</td>\n",
       "      <td>22.62</td>\n",
       "      <td>15.81</td>\n",
       "      <td>15.19</td>\n",
       "      <td>...</td>\n",
       "      <td>10.08</td>\n",
       "      <td>26.55</td>\n",
       "      <td>9.51</td>\n",
       "      <td>18.81</td>\n",
       "      <td>14.42</td>\n",
       "      <td>20.05</td>\n",
       "      <td>12.76</td>\n",
       "      <td>25.74</td>\n",
       "      <td>14.73</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2</td>\n",
       "      <td>33141200</td>\n",
       "      <td>33141300</td>\n",
       "      <td>10061.92</td>\n",
       "      <td>18901.03</td>\n",
       "      <td>662.90</td>\n",
       "      <td>476.56</td>\n",
       "      <td>12624.36</td>\n",
       "      <td>12815.05</td>\n",
       "      <td>8621.38</td>\n",
       "      <td>...</td>\n",
       "      <td>13262.95</td>\n",
       "      <td>617.76</td>\n",
       "      <td>454.16</td>\n",
       "      <td>493.94</td>\n",
       "      <td>15653.00</td>\n",
       "      <td>440.86</td>\n",
       "      <td>395.86</td>\n",
       "      <td>9882.05</td>\n",
       "      <td>408.98</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2</td>\n",
       "      <td>33141300</td>\n",
       "      <td>33141400</td>\n",
       "      <td>2922557.42</td>\n",
       "      <td>5869068.87</td>\n",
       "      <td>66513.92</td>\n",
       "      <td>47291.85</td>\n",
       "      <td>3747546.08</td>\n",
       "      <td>3914055.12</td>\n",
       "      <td>2341136.90</td>\n",
       "      <td>...</td>\n",
       "      <td>4048134.33</td>\n",
       "      <td>59262.13</td>\n",
       "      <td>45194.47</td>\n",
       "      <td>49687.68</td>\n",
       "      <td>4742511.44</td>\n",
       "      <td>46518.35</td>\n",
       "      <td>38207.37</td>\n",
       "      <td>2901473.37</td>\n",
       "      <td>40231.16</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2</td>\n",
       "      <td>33141400</td>\n",
       "      <td>33141500</td>\n",
       "      <td>2308283.28</td>\n",
       "      <td>4532227.15</td>\n",
       "      <td>71166.99</td>\n",
       "      <td>55250.19</td>\n",
       "      <td>3082463.35</td>\n",
       "      <td>2926391.11</td>\n",
       "      <td>2285220.69</td>\n",
       "      <td>...</td>\n",
       "      <td>3215387.23</td>\n",
       "      <td>69190.25</td>\n",
       "      <td>46811.78</td>\n",
       "      <td>52499.79</td>\n",
       "      <td>3897396.19</td>\n",
       "      <td>57191.96</td>\n",
       "      <td>39426.46</td>\n",
       "      <td>2365816.21</td>\n",
       "      <td>41335.07</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2</td>\n",
       "      <td>33141500</td>\n",
       "      <td>33141600</td>\n",
       "      <td>5895586.12</td>\n",
       "      <td>11218106.59</td>\n",
       "      <td>148116.46</td>\n",
       "      <td>113737.55</td>\n",
       "      <td>7746268.78</td>\n",
       "      <td>7473889.92</td>\n",
       "      <td>5767356.35</td>\n",
       "      <td>...</td>\n",
       "      <td>7946781.27</td>\n",
       "      <td>144794.17</td>\n",
       "      <td>95392.92</td>\n",
       "      <td>109254.27</td>\n",
       "      <td>9374017.79</td>\n",
       "      <td>113462.85</td>\n",
       "      <td>78539.73</td>\n",
       "      <td>6094810.55</td>\n",
       "      <td>82335.65</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2</td>\n",
       "      <td>33141600</td>\n",
       "      <td>33141700</td>\n",
       "      <td>3188920.02</td>\n",
       "      <td>6199340.58</td>\n",
       "      <td>53445.20</td>\n",
       "      <td>41923.58</td>\n",
       "      <td>4140398.21</td>\n",
       "      <td>4108340.56</td>\n",
       "      <td>2875539.71</td>\n",
       "      <td>...</td>\n",
       "      <td>4332996.57</td>\n",
       "      <td>51917.16</td>\n",
       "      <td>34638.30</td>\n",
       "      <td>39518.79</td>\n",
       "      <td>5090919.61</td>\n",
       "      <td>39525.65</td>\n",
       "      <td>28274.88</td>\n",
       "      <td>3227658.74</td>\n",
       "      <td>29720.33</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2</td>\n",
       "      <td>33141700</td>\n",
       "      <td>33141800</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.37</td>\n",
       "      <td>0.04</td>\n",
       "      <td>...</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2</td>\n",
       "      <td>33141800</td>\n",
       "      <td>33141900</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2</td>\n",
       "      <td>33141900</td>\n",
       "      <td>33142000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------------+-------------+-------+\n",
       "| Chromosome   | Start     | End       | LU-FF80     | LU-IO18     | +59   |\n",
       "| (category)   | (int32)   | (int32)   | (float64)   | (float64)   | ...   |\n",
       "|--------------+-----------+-----------+-------------+-------------+-------|\n",
       "| 2            | 33140000  | 33140100  | 42.0        | 39.76       | ...   |\n",
       "| 2            | 33140100  | 33140200  | 44.61       | 30.61       | ...   |\n",
       "| 2            | 33140200  | 33140300  | 47.35       | 38.67       | ...   |\n",
       "| 2            | 33140300  | 33140400  | 44.05       | 38.27       | ...   |\n",
       "| ...          | ...       | ...       | ...         | ...         | ...   |\n",
       "| 2            | 33141600  | 33141700  | 3188920.02  | 6199340.58  | ...   |\n",
       "| 2            | 33141700  | 33141800  | 0.19        | 0.14        | ...   |\n",
       "| 2            | 33141800  | 33141900  | 0.0         | 0.0         | ...   |\n",
       "| 2            | 33141900  | 33142000  | 0.0         | 0.0         | ...   |\n",
       "+--------------+-----------+-----------+-------------+-------------+-------+\n",
       "Unstranded PyRanges object has 20 rows and 64 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "59 hidden columns: LU-FF76, LU-14, LU-IO05, LU-IO23, LU-SC134, LU-FF4, LU-FF67, ... (+ 52 more.)"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_dfs_annot_gr['2', 33_140_000:33_142_000]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "genome_v7",
   "language": "python",
   "name": "genome_v7"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
